File: lib/feature/data/models/status.dart

import 'package:equatable/equatable.dart';

enum StatusEnum{
  initial,
  loading,
  successfully,
  failed,
}

class Status<T> extends Equatable{
  final StatusEnum status;
  final T? data;
  final String? message;

  const Status({
    required this.status,
    this.data,
    this.message
  });

  factory Status.initial(){
    return const Status(
        status: StatusEnum.initial
    );
  }

  factory Status.loading({T? data}){
    return Status(
      status: StatusEnum.loading,
      data: data,
    );
  }

  factory Status.success({required T? data, String? message}){
    return Status(
        status: StatusEnum.successfully,
        data: data,
        message: message ?? (data == null ? "Not found value" : "Success"),
    );
  }

  factory Status.failed({T? data, required String message}){
    return Status(
        status: StatusEnum.failed,
        data: data,
        message: message,
    );
  }

  @override
  List<Object?> get props => [status, data, message];
}

File: android/app/src/main/kotlin/com/example/spotify/data/model/status.kt
package com.example.spotify.data.model
import java.io.Serializable

class Status<out T> (val status: StatusEnum, val data: T?, val message: String?): Serializable {

    companion object{
        fun <T> Initialization(): Status<T> = Status<T>(StatusEnum.INITIALIZATION, null, null)

        fun <T> Loading(data: T?): Status<T> = Status<T>(StatusEnum.LOADING, data, null)

        fun <T> Success(data: T?): Status<T> = Status<T>(StatusEnum.SUCCESS, data, null)

        fun <T> Error(data: T?, message: String?): Status<T> = Status<T>(StatusEnum.ERROR, data, message)
    }
}

enum class StatusEnum: Serializable {
    INITIALIZATION,
    SUCCESS,
    ERROR,
    LOADING
}
